Linux/Mac: Log producer metadata information

gcmoreira opened the pull request:
This PR adds ISF metadata to log lines, including the producer name, version, types, and symbol sources for Mac and Linux. This makes it easier to understand the user's setup when they report an issue, saving time in troubleshooting. 
The information will be available at the `debug` log level, so `-vv`, `-vvv` and so on.

## Example 1 - Current GitHub ISF
This is the current `linux-sample-1.bin` ISF we are using for testing on GitHub, generated by `dwarf2json 0.9.0`. The types and symbols were extracted from `vmlinux-3.2.0-4-amd64`.

```shell
$ ./vol.py -vvv \
    -f ./linux-sample-1.bin \
    linux.pslist
Volatility 3 Framework 2.12.0
...
DEBUG    volatility3.framework.automagic.symbol_finder: Identified banner: b'Linux version 3.2.0-4-amd64 (debian-kernel@lists.debian.org) (gcc version 4.6.3 (Debian 4.6.3-14) ) #1 SMP Debian 3.2.57-3+deb7u2\n\x00'
DEBUG    volatility3.framework.automagic.symbol_finder: Using symbol library: file:///home/gmoreira/vol3_profiles/linux-image-3.2.0-4-amd64_3.2.57-3_deb7u2.json.xz
DEBUG    volatility3.framework.automagic.symbol_finder: producer_name: dwarf2json, producer_version: 0.9.0
DEBUG    volatility3.framework.automagic.symbol_finder: Types:
DEBUG    volatility3.framework.automagic.symbol_finder:         {'kind': 'dwarf', 'name': 'vmlinux-3.2.0-4-amd64', 'hash_type': 'sha256', 'hash_value': '67f9060253c801bffd3c99615c41d0ec266f8284c755f5368dad95b021481962'}
DEBUG    volatility3.framework.automagic.symbol_finder: Symbols:
DEBUG    volatility3.framework.automagic.symbol_finder:         {'kind': 'dwarf', 'name': 'vmlinux-3.2.0-4-amd64', 'hash_type': 'sha256', 'hash_value': '67f9060253c801bffd3c99615c41d0ec266f8284c755f5368dad95b021481962'}
DEBUG    volatility3.framework.automagic.symbol_finder:         {'kind': 'symtab', 'name': 'vmlinux-3.2.0-4-amd64', 'hash_type': 'sha256', 'hash_value': '67f9060253c801bffd3c99615c41d0ec266f8284c755f5368dad95b021481962'}
```

## Example 2 - Previous GitHub ISF
The following is the output from the ISF we previously used on GitHub. It was generated with `dwarf2json 0.6.0`, with types extracted from `vmlinux-3.2.0-4-amd64` and symbols sourced from `System.map-3.2.0-4-amd64` and `vmlinux-3.2.0-4-amd64`.

```shell
$ ./vol.py -vvv \
    -f ./linux-sample-1.bin \
    linux.pslist
Volatility 3 Framework 2.12.0
...
DEBUG    volatility3.framework.automagic.symbol_finder: Identified banner: b'Linux version 3.2.0-4-amd64 (debian-kernel@lists.debian.org) (gcc version 4.6.3 (Debian 4.6.3-14) ) #1 SMP Debian 3.2.57-3+deb7u2\n\x00'
DEBUG    volatility3.framework.automagic.symbol_finder: Using symbol library: file:///home/gmoreira/vol3_profiles/linux-3.2.0-4-amd64-debian.json.xz
INFO     volatility3.schemas: Dependency for validation unavailable: jsonschema
DEBUG    volatility3.schemas: All validations will report success, even with malformed input
DEBUG    volatility3.framework.automagic.symbol_finder: producer_name: dwarf2json, producer_version: 0.6.0
DEBUG    volatility3.framework.automagic.symbol_finder: Types:
DEBUG    volatility3.framework.automagic.symbol_finder:         {'kind': 'dwarf', 'name': 'vmlinux-3.2.0-4-amd64', 'hash_type': 'sha256', 'hash_value': 'b0b6a369b37c3a4764e6c0f8a57e1336490ba6c33bb1d83ca14e253a48ea359a'}
DEBUG    volatility3.framework.automagic.symbol_finder: Symbols:
DEBUG    volatility3.framework.automagic.symbol_finder:         {'kind': 'dwarf', 'name': 'vmlinux-3.2.0-4-amd64', 'hash_type': 'sha256', 'hash_value': 'b0b6a369b37c3a4764e6c0f8a57e1336490ba6c33bb1d83ca14e253a48ea359a'}
DEBUG    volatility3.framework.automagic.symbol_finder:         {'kind': 'symtab', 'name': 'vmlinux-3.2.0-4-amd64', 'hash_type': 'sha256', 'hash_value': 'b0b6a369b37c3a4764e6c0f8a57e1336490ba6c33bb1d83ca14e253a48ea359a'}
DEBUG    volatility3.framework.automagic.symbol_finder:         {'kind': 'system-map', 'name': 'System.map-3.2.0-4-amd64', 'hash_type': 'sha256', 'hash_value': 'a91a35cb30b78aa625e2a1e41a1b169c290b1e224c249c461caf9c34291791df'}
```

## Example 3 - Custom ISF
Here is a Volatility3 log from an ISF created with `dwarf2json 0.7.0`, using symbols from `System.map` and types extracted from a `module.ko` file.
```shell
$ ./vol.py -vvv \
    -f ./ubuntu120464bit_3.13.0-32_test_modules.core \
    linux.pslist
Volatility 3 Framework 2.12.0
...
DEBUG    volatility3.framework.automagic.symbol_finder: Identified banner: b'Linux version 3.13.0-32-generic (buildd@phianna) (gcc version 4.6.3 (Ubuntu/Linaro 4.6.3-1ubuntu5) ) #57~precise1-Ubuntu SMP Tue Jul 15 03:51:20 UTC 2014 (Ubuntu 3.13.0-32.57~precise1-generic 3.13.11.4)\n'
DEBUG    volatility3.framework.automagic.symbol_finder: Using symbol library: file:///home/gmoreira/vol3_profiles/Ubuntu_3.13.0-32.57~precise1_3.13.0-32-generic_x64.json
DEBUG    volatility3.framework.automagic.symbol_finder: producer_name: dwarf2json, producer_version: 0.7.0
DEBUG    volatility3.framework.automagic.symbol_finder: Types:
DEBUG    volatility3.framework.automagic.symbol_finder:         {'kind': 'dwarf', 'name': 'module.ko', 'hash_type': 'sha256', 'hash_value': 'de6c4c45a919def20b2991496b9d964f57dae212f3ba70778ca3e7f0f1a8c6f5'}
DEBUG    volatility3.framework.automagic.symbol_finder: Symbols:
DEBUG    volatility3.framework.automagic.symbol_finder:         {'kind': 'system-map', 'name': 'System.map', 'hash_type': 'sha256', 'hash_value': '0590546715deed2ee83865b3117a991373adf18e4a48ba679d8e108484c74554'}
```

# Mac
The changes include support for Mac. The Mac ISF json format is similar to Linux, so it should work seamlessly. However, I currently don't have a Mac sample on my machine, so I would greatly appreciate it if someone else could test this for me.

gcmoreira committed:
Linux/Mac: Log producer information

ikelos commented on volatility3/framework/automagic/symbol_finder.py:
```diff
@@ -160,8 +160,29 @@ def _banner_scan(
+                    vollog.debug(
+                        f"producer_name: {producer.name}, producer_version: {producer.version_string}"
+                    )
+                    for category in symbol_table.metadata._json_data:
```
We shouldn't be accessing private members.  It looks like you already added support for returning an appropriate `Metadata` object for Mac.  If there's specific things you need from the linux/mac metadata, they should be exposed by the various `*Metadata` classes.

They also won't have been part of the schema, so you'll need to write them defensively to return no data if they're not populated, but not to fail (and possibly to update the schema, so things like `dwarf2json` or `btf2json` know that they should populate them in future)...

gcmoreira replied:
done

ikelos reviewed (commented):
Generally looks good and thorough and I definitely like the concept, but digging into the metadata and pulling out the private JSON isn't cool, there needs to be specific methods/properties for doing that in case we one day pivot away from JSON as the storage format....

gcmoreira committed:
Linux/Mac: Address code review suggestions

- Add getters for Linux/Mac ISF sources
- Avoid using internal attributes
- Use the dict repr instead of walking the dict to simplify code

gcmoreira commented:
Hey @ikelos thanks for your review! I've made the changes, please take a look and let me know if everything looks good now

gcmoreira committed:
Linux/Mac: ISF metadata: Rename s/DWARF/POSIX/, as I'm not happy with the generic name. BTF source could potentially generate the same keys

ikelos commented on volatility3/framework/symbols/metadata.py:
```diff
@@ -86,9 +85,21 @@ def pdb_age(self) -> Optional[int]:
 
 
-class LinuxMetadata(interfaces.symbols.MetadataInterface):
+class DwarfMetadata(interfaces.symbols.MetadataInterface):
```
Is this specifically DWARF based?  I'm just thinking if we've got btf2json in the pipeline, will it look the same or not?

gcmoreira replied:
I know, I later changed that. I think you reviewed an older version of it? See my comment on 77778ee

ikelos reviewed (approved):
Thanks for the fix up, looks much better (you even caught some `_dict` accesses I hadn't noticed)!  5:P

Just a question remaining but otherwise looks done.  5:)

ikelos commented:
Ah yep, sorry.  I was trying to be clever and look at just the stuff that had changed since last time, but I only saw the one commit...  Cool, looks good then!  5:D
